### DISCUSSION TITLE
Problem with expression result (unexpected result)

### DISCUSSION THREAD

**Discussion by Megabyte94** [ezylang/EvalEx#296] [Q&A]
Good evening.
I am having a probem with the result of an expression that EvalEX does return...

Specifically, I'm trying to make an Android Toast appear and show the result of an expression by this code:
`Expression expression = new Expression("2*" + String.valueOf(Math.PI) + "*180/" + String.valueOf(Math.PI));
expression.setPrecision(13);
BigDecimal result = expression.eval();
Toast.makeText(mContext, String.valueOf(result), Toast.LENGTH_SHORT).show();`

But it is showing "**359.9999999999**" as the result, instead of "**360**". Why is this problem occurring? 
I want the expression to always have maximum 13 decimal values. I know I can approximate the BigDecimal result to an Integer and get "360" as the result, but I don't want to do this, I want to get a BigDecimal result with maximum 13 decimal numbers with any expression I try to calculate using this code and this expression to return "360" as the correct result. But I don't understand why is EvalEX implementation showing me this strange result...

I would be very glad if you could help me to find a solution to this problem I'm currently having. Thanks.

**Comment by RobertZenz** [comment]
First, you're not precise enough. You should use the builtin `PI` constant instead of embedding a **`double` converted to a `String`** in the expression:

```java
Expression expression = new Expression("2*PI*180/PI")
```

Second, that's not how the precision on the `BigDecimal` works. Precision sets the number of exact digits, meaning that the number "123.45" has a precision of 5, but a *scale* of 2. You want to set precision to something big, and then scale the `BigDecimal` to the appropriate number of decimal places.

**Reply by Megabyte94** [reply]
Thank you so much for your time and help.

Yeah, I just realized that number precision indicates the maximum of total digits and not only decimal digits, thanks again.
But I tried using the built in "PI" constant instead of the String value of "Math.PI", and I am having the same result... The Toast String is exactly the same... How could I solve this?

This is the new code that now uses the constant "PI" in the expression but still throwing the same Toast String:
`Expression expression = new Expression("2*PI*180/PI");
 expression.setPrecision(13);
 BigDecimal result = expression.eval();
 Toast.makeText(mContext, String.valueOf(result), Toast.LENGTH_SHORT).show();`

**Reply by RobertZenz** [reply]
My best guess is that we're looking at a rounding issue. If we try every precision from 1 to 20 we get the following results:

```
 1: 300
 2: 350
 3: 360
 4: 360
 5: 360.01
 6: 359.999
 7: 359.9998
 8: 360.00001
 9: 359.999999
10: 359.9999999
11: 360
12: 359.999999999
13: 359.9999999999
14: 359.99999999999
15: 359.999999999999
16: 359.9999999999998
17: 360.00000000000002
18: 360.000000000000001
19: 360.0000000000000001
20: 360.00000000000000001
```

I could now work this out by hand, but to be honest I don't feel like it. We are still dealing with limited precision to a certain degree even though we are using `BigDecimal` and PI is quite a large number with no end in sight. A similar issue is the expression `1/(1/60)`, which should yield `1` as result, but simply doesn't `59.999999999999999999`. It's a problem when dealing with limited precision.

As a workaround one can always do the calculation with a larger precision and then round to the wanted precision afterwards:

```java
Expression expression = new Expression("2*PI*180/PI");
expression.setPrecision(13 * 2);
BigDecimal result = expression.eval();
BigDecimal resultWithWantedPrevision = result.round(new MathContext(13));

System.out.println(resultWithWantedPrevision.toPlainString());
```

**Comment by uklimaschewski** [comment]
I did a quick test with EvalEx 3 and there the expression `2*PI*180/PI` evaluates to 360:
```Java
  @Test
  void testPi() throws EvaluationException, ParseException {
    Expression expression = new Expression("2*PI*180/PI");

    assertThat(expression.evaluate().getStringValue()).isEqualTo("360");
  }
```
The reason is that version 3 has a different default math context. It has a higher precision (68 instead of 7) using the same rounding mode (HALF_EVEN).
If I use that math context with version 2, I get the same result:
```Java
  @Test
  public void testPi() {
    Expression expression = new Expression("2*PI*180/PI",
        new MathContext(68, RoundingMode.HALF_EVEN));

    assertEquals(expression.eval().toPlainString(), "360");
  }
```
